cmd/server.go

```diff
@@ -195,15 +195,23 @@ func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
 					}
 					if err != nil {
 						if etcdErr, ok := err.(*errors.EtcdError); ok == true {
-							logger.Errorf("Failed to take first delta snapshot: snapshotter failed with etcd error: %v", etcdErr)
+							logger.Warnf("Failed to take first delta snapshot: snapshotter failed with etcd error: %v", etcdErr)
 						} else {
-							logger.Errorf("Failed to take first delta snapshot: snapshotter failed with error: %v", err)
+							logger.Warnf("Failed to take first delta snapshot: snapshotter failed with error: %v", err)
+						}
+						startWithFullSnapshot = false
+						if err := ssr.TakeFullSnapshotAndResetTimer(); err != nil {
+							logger.Errorf("Failed to take substitute first full snapshot: %v", err)
+							continue
 						}
-						continue
 					}
 					if err = ssr.TakeDeltaSnapshot(); err != nil {
-						logger.Errorf("Failed to take first delta snapshot: snapshotter failed with error: %v", err)
-						continue
+						logger.Warnf("Failed to take first delta snapshot: snapshotter failed with error: %v", err)
+						startWithFullSnapshot = false
+						if err := ssr.TakeFullSnapshotAndResetTimer(); err != nil {
```

Comment:
If taking delta snapshot failed means, there is issue with snapshot upload. And ssr.TakeFullSnapshotAndResetTimer() will also fail with same reason. So, it's better to fail here and retry for delta snapshot.

Reply:
Not necessarily. The primary motivation for this PR was to handle cases where the watch revision is not present in etcd due to compaction before backup sidecar starts. If delta snapshot fails due to a bucket error or network issue, then even on the next attempt there's a good chance the delta snapshot upload will fail. So I see no difference between retrying a delta snapshot and trying a full snapshot, because anyway full snapshot is done as chunk upload, so there's no issue of eating up network bandwidth as well.

Reply:
Yes. but the time difference is more with full snapshot. That's the main point behind having this optimisation with first delta snapshot. On failure to take initial delta snapshot due cloud provider/network error, situation boil down to same point before etcd-backup-restore restart.

Reply:
Ah ok. There's one thing I can do. We know that the revision compaction issue is actually encountered by the `CollectEventsSincePrevSnapshot` call. If this call fails, we can attempt a full snapshot (because this call has nothing to do with the bucket itself, so provider/network errors won't occur here). But if it's an error with `TakeDeltaSnapshot`, then we re-attempt delta snapshot. WDYT?

Reply:
Yup, by re-attempt means - continue with the for loop; don't call TakeDeltaSnapshot again.

Reply:
Yes, by re-attempt delta snapshot even I meant re-enter the loop.  :)